### ISSUE TITLE
Safety level may not be changed inside a transaction: Circleci build fail with Coverage 5.0a2

### ISSUE THREAD

**Issue by gbolly** [coveragepy/coveragepy#703]
I was running circleci build using tox to run against python version 3.6 and all my tests seems to be passing but crashed. This was the error message I got:

```Couldn't use data file '/root/my-package/.coverage': Safety level may not be changed inside a transaction
ERROR: InvocationError for command '/root/my-package/.tox/py36/bin/coverage run setup.py test' (exited with code 1)
___________________________________ summary ____________________________________
ERROR:   py36: commands failed```

Failed CI build: https://circleci.com/gh/datadotworld/tap-redshift/78

I had also experienced this on another project i'm working on. Any pointers?

**Comment by nedbat** [comment]
The error on CircleCI is "Safety level may not be changed inside a transaction".  I don't understand why that would happen to you when it doesn't happen to others.  Can you run the test again with the environment variable COVERAGE_DEBUG=dataop,dataio,sql ?

**Comment by hugoduncan** [comment]
Coverage.py, version 5.0a3 with C extension

```
COVERAGE_DEBUG=dataop,dataio,sql pytest -v
Erasing data file '/Users/duncan/projects/local/myproj/.coverage'
Setting context: None
Erasing data file '/Users/duncan/projects/local/myproj/.coverage'
Creating data file '/Users/duncan/projects/local/myproj/.coverage'
Connecting to '/Users/duncan/projects/local/myproj/.coverage'
Executing 'pragma journal_mode=off'
Executing 'pragma synchronous=off'
Traceback (most recent call last):
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/coverage/sqldata.py", line 532, in execute
    return self.con.execute(sql, parameters)
sqlite3.OperationalError: Safety level may not be changed inside a transaction

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/bin/pytest", line 11, in <module>
    sys.exit(main())
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/_pytest/config/__init__.py", line 58, in main
    config = _prepareconfig(args, plugins)
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/_pytest/config/__init__.py", line 196, in _prepareconfig
    pluginmanager=pluginmanager, args=args
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/pluggy/hooks.py", line 284, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/pluggy/manager.py", line 67, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/pluggy/manager.py", line 61, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/pluggy/callers.py", line 203, in _multicall
    gen.send(outcome)
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/_pytest/helpconfig.py", line 93, in pytest_cmdline_parse
    config = outcome.get_result()
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/pluggy/callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/_pytest/config/__init__.py", line 675, in pytest_cmdline_parse
    self.parse(args)
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/_pytest/config/__init__.py", line 845, in parse
    self._preparse(args, addopts=addopts)
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/_pytest/config/__init__.py", line 809, in _preparse
    early_config=self, args=args, parser=self._parser
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/pluggy/hooks.py", line 284, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/pluggy/manager.py", line 67, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/pluggy/manager.py", line 61, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/pluggy/callers.py", line 208, in _multicall
    return outcome.get_result()
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/pluggy/callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/pytest_cov/plugin.py", line 98, in pytest_load_initial_conftests
    plugin = CovPlugin(early_config.known_args_namespace, early_config.pluginmanager)
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/pytest_cov/plugin.py", line 145, in __init__
    self.start(engine.Central)
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/pytest_cov/plugin.py", line 167, in start
    self.cov_controller.start()
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/pytest_cov/engine.py", line 160, in start
    self.cov.start()
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/coverage/control.py", line 431, in start
    self._init_for_start()
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/coverage/control.py", line 380, in _init_for_start
    self._collector.use_data(self._data, self.config.context)
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/coverage/collector.py", line 167, in use_data
    self.covdata.set_context(self.static_context)
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/coverage/sqldata.py", line 210, in set_context
    with self._connect() as con:
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/coverage/sqldata.py", line 159, in _connect
    self._create_db()
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/coverage/sqldata.py", line 116, in _create_db
    with self._db:
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/coverage/sqldata.py", line 516, in __enter__
    self.connect()
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/coverage/sqldata.py", line 509, in connect
    self.execute("pragma synchronous=off").close()
  File "/Users/duncan/.local/share/virtualenvs/myproj-EL7_EUmm/lib/python3.6/site-packages/coverage/sqldata.py", line 534, in execute
    raise CoverageException("Couldn't use data file {!r}: {}".format(self.filename, exc))
coverage.misc.CoverageException: Couldn't use data file '/Users/duncan/projects/local/myproj/.coverage': Safety level may not be changed inside a transaction
```

**Comment by strichter** [comment]
@gbolly We have seen several of these type of errors too around this code path, because it is not properly protected. It happens due to race conditions when using multiple threads.

Try applying this PR: https://github.com/nedbat/coveragepy/pull/760

**Comment by nedbat** [comment]
@gbolly @hugoduncan Have you seen this problem with more recent 5.0 alpha versions?

**Comment by nedbat** [comment]
I haven't heard about this in a long time.  Please re-open if it happens again.

**Comment by drolando** [comment]
@nedbat I just got the same error running https://github.com/Yelp/py_zipkin tests. Interestingly it only happens for python 3.6 on ubuntu xenial (python==3.6.0), works fine on py27, py37 and pypy.
It also works fine on python 3.6.9, so there's something special about that specific python version.

**Comment by hardbyte** [comment]
Also seem to be running into this specifically on Python 3.6.0 on MACOS builds on travis-ci. 

(I can't reopen as not a `coveragepy` contributor)

**Comment by nedbat** [comment]
@hardbyte Can you give me a way to reproduce it? Have you tried with a more recent 3.6.x version of Python?

**Comment by mgedmin** [comment]
I'm seeing this too!  Python 3.6.0 on MacOS on Travis CI:
- https://travis-ci.org/zopefoundation/persistent/jobs/637976683
- https://travis-ci.org/zopefoundation/persistent/jobs/637976872

**Comment by nedbat** [comment]
Seems like the reports are all for 3.6.0 specifically.  Is it happening on any other version of Python?

**Comment by mgedmin** [comment]
Experiments show that upgrading to Python 3.6.1 makes this problem go away.

Maybe it was ~~caused~~ fixed by https://bugs.python.org/issue28518?  It's the only sqlite-related change I can find between CPython v3.6.0 and v3.6.1.

**Comment by mgedmin** [comment]
Yeah, it's that bug.  A [comment](https://bugs.python.org/msg284585) calls out `PRAGMA synchronous` explicitly.

Before https://github.com/python/cpython/commit/ab994ed8b97e1b0dac151ec827c857f5e7277565 (released in v3.6.0b1) Python's sqlite module would do an implicit autocommit before executing the PRAGMA.  From that commit  you'd get this error if you don't do an explicit commit().  Eventually it was all fixed in https://github.com/python/cpython/commit/76995cab69d5ef83d31d8a5754cbad11be4038cb (released in v3.6.1rc1).

**Comment by nedbat** [comment]
Thanks for finding it.  We can close this again.
